File: ValidSudoku/same.java
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        if(board == null ||board.length != 9 || board[0].length != 9) return false;
        Set<Character> rowSet = new HashSet<Character>();
        Set<Character> colSet = new HashSet<Character>();
        for(int i = 0;i< 9;i++){
            for(int j = 0;j < 9;j++){
                // 验证每一行是否满足数独条件
                if(board[i][j] != '.'){
                    if(!rowSet.add(board[i][j])){
                        return false;
                    }
                }
                //验证每一列是否满足数独条件
                if(board[j][i] !='.'){
                    if(!colSet.add(board[j][i])){
                      return false;
                    }
                }
            }
            rowSet.clear();
            colSet.clear();
        }
        rowSet.clear();
        //验证每个9宫格是否满足数独条件
        for(int i = 0;i< 3;i++){
            for(int j = 0;j<3;j++){
                for(int m = 3*i; m < (3 +3*i); m++){
                    for(int n = 3*j; n < (3 + 3*j); n++){
                        if(board[m][n] != '.'){
                            if(!rowSet.add(board[m][n])){
                                return false;
                            }
                        }
                    } 
                }
                rowSet.clear();
            }
        }
        return true;
    }
}

File: ValidSudoku/same.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for (int i=0;i<9;i++){
            for (int j=0;j<9;j++){
                if (board[i][j]=='.'){
                    continue;
                }
                for (int k=0;k<9;k++){
                    if (j<k&&board[i][j]==board[i][k]){
                        return false;
                    }
                    if (i<k&&board[i][j]==board[k][j]){
                        return false;
                    }
                    if (board[i][j]==board[i-i%3+k/3][j-j%3+k%3]){
                        if (i-i%3+k/3!=i||j-j%3+k%3!=j){
                            return false;
                        }
                    }
                }
            }
        }
        return true;
    }
};
